### ISSUE TITLE
"next" statement called as a coroutine in case %autoawait is on

### ISSUE THREAD

**Issue by DrJackilD** [ipython/ipython#12693]
Hi guys,

I found one strange behavior when I run a simple next statement in ipython interactive console with `autoawait` on.

Example:
```shell
In[1]: next((i for i in range(5) if i == 6))
Traceback (most recent call last):
  File "<ipython-input-1-ab0f6ef39cbd>", line 1, in <module>
    next((i for i in range(5) if i == 6))
StopIteration
The above exception was the direct cause of the following exception:
Traceback (most recent call last):
  File "/usr/local/anaconda3/envs/allset_menu/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3341, in run_code
    await eval(code_obj, self.user_global_ns, self.user_ns)
RuntimeError: coroutine raised StopIteration
```

Instead of StopIteration, I've got a RuntimeError. It's appeared not in all ipython's shells, only if it's got some Python versions.

My environment:
Python version: 3.8.3
ipython version: 7.16.1

Also, I tried to reproduce this error in other environments, it didn't work. I've checked Python 3.8.1 and Python 3.7.5, nothing happened, I've got StopIteration as planned. I made some debug and found a place, where this can appear:

```python
# IPython.core.interactiveshell.InteractiveShell.run_ast_nodes
3239:           if sys.version_info > (3, 8):
                    def compare(code):
                        is_async = (inspect.CO_COROUTINE & code.co_flags == inspect.CO_COROUTINE)
                        return is_async
                else:
                    def compare(code):
                        return _async
```

I believe the problem is here, when we going to check the flags in an AST object
`(inspect.CO_COROUTINE & code.co_flags == inspect.CO_COROUTINE)`
But of course I can be wrong.

**Comment by DrJackilD** [comment]
I will do a PR myself, just need to point out am I correct with my assumptions. :)

**Comment by Carreau** [comment]
No that's a but in CPython that have been fixed in I believe the 3.8.4 patch release. See #12422 

See upstream bug: https://bugs.python.org/issue41218

It aggressively mark things as coroutine when they are not. AFAICT there is no proper way to fix it in IPython itself.

**Comment by DrJackilD** [comment]
@Carreau thank you for the explanation! I'll close this ticket now.
